File: backend/src/routes/analytics.js
import express from 'express';
const router = express.Router();
import { auth } from '../middleware/auth.js';
import Task from '../models/Task.js';
import ProgressLog from '../models/ProgressLog.js';
import User from '../models/User.js';

router.get('/overview', auth, async (req, res) => {
  try {
    const { range = 'week' } = req.query;
    
    const tasks = await Task.find({ archived: false });
    const progress = await ProgressLog.find();
    const users = await User.find({ role: 'collaborator', active: true });

    const now = new Date();
    const weekAgo = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000);
    const recentProgress = progress.filter(p => new Date(p.createdAt) > weekAgo);

    const weeklyHoursData = [];
    for (let i = 6; i >= 0; i--) {
      const date = new Date(now.getTime() - i * 24 * 60 * 60 * 1000);
      const dayProgress = progress.filter(p => {
        const pDate = new Date(p.createdAt);
        return pDate.toDateString() === date.toDateString();
      });
      const hours = dayProgress.reduce((sum, p) => sum + p.hoursSpent, 0);
      weeklyHoursData.push({
        date: date.toLocaleDateString('en-US', { weekday: 'short' }),
        hours: Math.round(hours * 10) / 10,
      });
    }

    const taskStatusDistribution = [
      {
        status: 'To Do',
        count: tasks.filter(t => t.status === 'todo').length,
        collaborators: [],
      },
      {
        status: 'In Progress',
        count: tasks.filter(t => t.status === 'in-progress').length,
        collaborators: [],
      },
      {
        status: 'Review',
        count: tasks.filter(t => t.status === 'review').length,
        collaborators: [],
      },
      {
        status: 'Done',
        count: tasks.filter(t => t.status === 'done').length,
        collaborators: [],
      },
      {
        status: 'Blocked',
        count: tasks.filter(t => t.status === 'blocked').length,
        collaborators: [],
      },
    ].filter(item => item.count > 0);

    for (const item of taskStatusDistribution) {
      const statusTasks = tasks.filter(t => {
        const statusMap = {
          'To Do': 'todo',
          'In Progress': 'in-progress',
          'Review': 'review',
          'Done': 'done',
          'Blocked': 'blocked',
        };
        return t.status === statusMap[item.status];
      });
      
      const collaboratorIds = new Set();
      statusTasks.forEach(t => {
        t.assignedTo.forEach(id => collaboratorIds.add(id.toString()));
      });
      
      const collaborators = await User.find({
        _id: { $in: Array.from(collaboratorIds) },
      });
      
      item.collaborators = collaborators.map(u => u.name);
    }

    const topContributors = await Promise.all(
      users.map(async (u) => {
        const userProgress = progress.filter(p => p.userId.toString() === u._id.toString());
        const hours = userProgress.reduce((sum, p) => sum + p.hoursSpent, 0);
        return {
          name: u.name,
          hours: Math.round(hours * 10) / 10,
        };
      })
    );

    topContributors.sort((a, b) => b.hours - a.hours);

    res.json({
      tasksCompleted: tasks.filter(t => t.status === 'done').length,
      activeContributors: users.length,
      hoursThisWeek: Math.round(recentProgress.reduce((sum, p) => sum + p.hoursSpent, 0) * 10) / 10,
      weeklyHoursData,
      taskStatusDistribution,
      topContributors: topContributors.slice(0, 5),
      overdueTasks: tasks.filter(t => {
        if (!t.deadline) return false;
        return new Date(t.deadline) < now && t.status !== 'done';
      }).length,
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

File: backend/app/routes/analytics.py
from datetime import datetime, timedelta, timezone
from fastapi import APIRouter, Depends, Query
from app.auth import get_current_user
from app.database import get_admin_client

router = APIRouter()


@router.get("/overview")
async def get_analytics_overview(
    time_range: str = Query("week", alias="range"),
    current_user: dict = Depends(get_current_user),
):
    supabase = get_admin_client()

    tasks_response = supabase.table("tasks").select("*").eq("archived", False).execute()
    progress_response = supabase.table("progress_logs").select("*").execute()
    users_response = (
        supabase.table("profiles")
        .select("*")
        .eq("role", "collaborator")
        .eq("active", True)
        .execute()
    )

    tasks = tasks_response.data
    progress = progress_response.data
    users = users_response.data

    now = datetime.now(timezone.utc)
    week_ago = now - timedelta(days=7)

    recent_progress = [
        p
        for p in progress
        if p.get("created_at")
        and datetime.fromisoformat(p["created_at"].replace("Z", "+00:00")) > week_ago
    ]

    hours_this_week = round(sum(p.get("hours_spent", 0) for p in recent_progress), 1)

    weekly_hours_data = []
    for i in range(6, -1, -1):
        date = now - timedelta(days=i)
        day_str = date.strftime("%a")
        day_progress = [
            p
            for p in progress
            if p.get("created_at")
            and datetime.fromisoformat(p["created_at"].replace("Z", "+00:00")).strftime(
                "%Y-%m-%d"
            )
            == date.strftime("%Y-%m-%d")
        ]
        hours = round(sum(p.get("hours_spent", 0) for p in day_progress), 1)
        weekly_hours_data.append({"date": day_str, "hours": hours})

    status_counts = {}
    for t in tasks:
        s = t.get("status", "todo")
        status_counts[s] = status_counts.get(s, 0) + 1

    status_labels = {
        "todo": "To Do",
        "in-progress": "In Progress",
        "review": "Review",
        "done": "Done",
        "blocked": "Blocked",
    }

    task_status_distribution = []
    for status_key, label in status_labels.items():
        if status_key in status_counts:
            status_tasks = [t for t in tasks if t.get("status") == status_key]
            collaborator_ids = set()
            for t in status_tasks:
                for uid in t.get("assigned_to", []):
                    collaborator_ids.add(uid)

            user_names = []
            for uid in collaborator_ids:
                user_match = [u for u in users if u["id"] == uid]
                if user_match:
                    user_names.append(user_match[0]["name"])

            task_status_distribution.append(
                {
                    "status": label,
                    "count": status_counts[status_key],
                    "collaborators": user_names,
                }
            )

    top_contributors = []
    for u in users:
        user_progress = [p for p in progress if p.get("user_id") == u["id"]]
        hours = round(sum(p.get("hours_spent", 0) for p in user_progress), 1)
        top_contributors.append({"name": u["name"], "hours": hours})

    top_contributors.sort(key=lambda x: x["hours"], reverse=True)

    overdue_tasks = sum(
        1
        for t in tasks
        if t.get("deadline")
        and datetime.fromisoformat(t["deadline"].replace("Z", "+00:00")) < now
        and t.get("status") != "done"
    )

    return {
        "tasksCompleted": status_counts.get("done", 0),
        "activeContributors": len(users),
        "hoursThisWeek": hours_this_week,
        "weeklyHoursData": weekly_hours_data,
        "taskStatusDistribution": task_status_distribution,
        "topContributors": top_contributors[:5],
        "overdueTasks": overdue_tasks,
    }
